### ISSUE TITLE
cli: Don’t create root-owned files in Docker

### ISSUE THREAD

**Issue by wtraylor** [getpopper/popper#859]
Files that are created by executing `docker run` are owned by user `root` and cannot easily be deleted.
# Steps to Reproduce
```bash
mkdir project
cd project
# Create a workflow that creates some cascaded folder structure:
cat >wf.yml <<EOF
steps:
- uses: "docker://alpine:latest"
  args: ["mkdir", "--parents", "testdir/subdir"]
EOF
popper run -f wf.yml

ls -l testdir
# drwxr-xr-x 2 root root 40 Jun 17 13:42 subdir/

# Try to delete the created folder:
rm --recursive --force testdir
# rm: cannot remove 'testdir/subdir': Permission denied
```
Note that calling `rm` on just one file/folder works with an interactive confirmation.

OS: Archlinux
Popper: Current `master` (b682359f06863935feaf9bac6678e583fb4caf9c) in Python 3.8.3
Docker: version 19.03.11-ce, build 42e35e61f3
# Expected Behavior
Files and folders created by executing Popper are owned by the user invoking Popper.

# Suggested Fix
Run Docker with the `--user` option so that all created files are owned by the user calling `popper run`:
```bash
docker run --user $(id -u):$(id -g) ...
```

**Comment by wtraylor** [comment]
Compare also [this blogpost](https://medium.com/redbubble/running-a-docker-container-as-a-non-root-user-7d2e00f8ee15) by Lucas Wilson-Richter, which suggests the `--user` flag and discusses implications.

**Comment by ivotron** [comment]
thanks a lot @wtraylor, really appreciate the detailed explanation and suggested fix!

Yes, this is indeed a VERY annoying issue when running on Linux (see https://github.com/moby/moby/issues/2259). Paradoxically, the user experience of running Docker (**Linux** Containers) on Mac or Windows is friendlier than on Linux (thanks to the filesystem sharing feature of the underlying hypervisor).

In addition to what that blogpost you linked mentions, an outcome of  doing `--user $(id -u):$(id -g)` is that it can only write to folders where the "others" group is allowed. This boils down to only being able to write to `/tmp`, so if the container's entrypoint (or the given command) writes a file to any place outside `/workspace` or `/tmp`, the corresponding step is going to fail. This wouldn't be an issue if it wasn't for the very common practice of container images assuming that the user running is `root`, or that a user gets created at image build-time that might not have the same UID than the user running `popper run`.

So to summarize, having `popper run` always include the `--user` flag when running a step solves the issue of not ending with files in the workspace that are owned by root, but results in potential errors that originate in programs writing to places other than `/workspace` or `/tmp`.

In an ideal scenario, a container would run everything as root inside the container, but then write to bind-mounted folders as the user on the host. This is precisely the problem that Podman (and rootless Docker) are trying to solve. So we can either wait for #704 to be fixed, or find a workaround that addresses the tradeoff described in the previous paragraph. For example, we could add a `run_as_user` boolean attribute to steps that instructs Popper to do `--user`. Another "hack" would be to run `chown` (from within a container) at the end of the workflow execution, so that all the contents of `/workspace` have the right user/group ownership.

what do you think?

**Comment by wtraylor** [comment]
Thank you for the explanations, @ivotron. [Podman](https://podman.io/) looks promising.

> Another "hack" would be to run chown (from within a container) at the end of the workflow execution, so that all the contents of /workspace have the right user/group ownership.
 
Calling `chown` on `/workspace` might cause unintended side effects. Perhaps the project contains files that are *supposed* to be owned by another user.
I don’t know how one could limit `chown` to *only* the files created by the containerized workflow.

> For example, we could add a `run_as_user` boolean attribute to steps that instructs Popper to do --user. 

The beauty of Popper lies in its simplicity, and I am reluctant to vote for adding new config variables. However, `run_as_user` would be a straight shot, and I would like this option if it had no problematic side effects.

Coming from a user’s perspective, my expectation is to assume this behavior in the first place. In my field (ecology), few people know how Docker works. My colleagues would just *expect* that everything created is owned by the calling user. So I would suggest to either introduce a flag `run_as_root` for those Popper users who need to be `root`, or add the `run_as_user` flag to the default Popper scaffold.

Now the issues: Running a container as a non-root user creates the problem that you cannot install packages in the container:
```bash
docker run --user (id -u):(id -g) --rm  debian apt install git
# E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
# E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
```
And `sudo` is usually not preinstalled, and would first require adding the user to `/etc/sudoers`.
```bash
docker run --user (id -u):(id -g) --rm  debian sudo apt install git
# docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"sudo\": executable file not found in $PATH": unknown.
```
Of course, you can create your custom Dockerfile for installing software, but inexperienced users will first try to put their install commands into their script.

In conclusion, I think non of the workarounds offers an easy solution for the average Popper user.

**Comment by ivotron** [comment]
> In conclusion, I think non of the workarounds offers an easy solution for the average Popper user.

I agree 100%. Given that @edeediong is already working on #704 , I think we can wait for that to be completed and hope that this issue gets addressed as well. Thanks @wtraylor! 🙏

**Comment by edeediong** [comment]
Ha! A first use case for Podman! I'm sure gonna complete the Podman feature soon.

**Comment by ivotron** [comment]
now that #903 has been merged, we can close this. Please feel free to re-open if you think the podman runner is not addressing this issue

**Comment by white-gecko** [comment]
I had the same issue, but was not able to run podman in this situation.
So I solved it by setting `engine.options.user:` in the engine configuration.
`popper run -c engine.yml …`

`engine.yml`:
```
engine:
  name: docker
  options:
     volumes:
     - /local/path:/path/in/docker
     user: "1000:1000"
```
or any other userid and groupid.

Should be possible to set the user and group per step as well.
